python/ql/consistency-queries/DataFlowConsistency.ql

```diff
@@ -0,0 +1,122 @@
+    //
+    // TODO: Figure out how bad breaking this consistency check is
+    exists(Function func, Parameter param |
+      c.getScope() = func and
+      p = parameterNode(param) and
+      c.getParameter(pos) = p and
+      param = func.getArg(_) and
+      param = func.getArgByName(_)
+    )
+  }
+
+  predicate uniqueCallEnclosingCallableExclude(DataFlowCall call) {
+    not exists(call.getLocation().getFile().getRelativePath())
+  }
+
+  predicate identityLocalStepExclude(Node n) {
+    not exists(n.getLocation().getFile().getRelativePath())
+  }
+
+  predicate multipleArgumentCallExclude(ArgumentNode arg, DataFlowCall call) {
+    // since we can have multiple DataFlowCall for a CallNode (for example if can
+    // resolve to multiple functions), but we only make _one_ ArgumentNode for each
+    // argument in the CallNode, we end up violating this consistency check in those
+    // cases. (see `getCallArg` in DataFlowDispatch.qll)
+    exists(DataFlowCall other, CallNode cfgCall | other != call |
+      call.getNode() = cfgCall and
+      other.getNode() = cfgCall and
+      isArgumentNode(arg, call, _) and
+      isArgumentNode(arg, other, _)
+    )
+    or
+    // bound methods that refer to the same self argument.
+    // Example: In `bm = self.foo; bm(); bm()` both bm() calls use the same `self` as
+    // the (pos self) argument
+    exists(AttrRead attr, DataFlowCall other | other != call |
+      // for simple cases we can track the function back to the attr read but when the
+      // call appears in the body of a list-comprehension, we can't do that, and simply
+      // allow it instead.
+      (
+        call.getScope() = attr.getScope() and
+        any(CfgNode n | n.asCfgNode() = call.getNode().(CallNode).getFunction()).getALocalSource() =
+          attr
+        or
+        not exists(call.getScope().(Function).getDefinition()) and
+        call.getScope().getScope+() = attr.getScope()
+      ) and
+      (
+        other.getScope() = attr.getScope() and
+        any(CfgNode n | n.asCfgNode() = other.getNode().(CallNode).getFunction()).getALocalSource() =
+          attr
+        or
+        not exists(other.getScope().(Function).getDefinition()) and
+        other.getScope().getScope+() = attr.getScope()
+      ) and
+      arg = attr.getObject() and
+      arg = call.getArgument(any(ArgumentPosition p | p.isSelf())) and
+      arg = other.getArgument(any(ArgumentPosition p | p.isSelf()))
+    )
+    or
+    // `f = getattr(obj, "foo"); f()` where `obj` is used as (pos self) argument for
+    // `f()` call
+    exists(DataFlowCall getAttrCall, DataFlowCall methodCall, AttrRead attr |
+      call in [getAttrCall, methodCall]
+    |
+      arg = getAttrCall.getArgument(any(ArgumentPosition p | p.isPositional(0))) and
+      arg = methodCall.getArgument(any(ArgumentPosition p | p.isSelf())) and
+      attr.getObject() = arg and
+      attr.(CfgNode).getNode() = getAttrCall.getNode()
+    )
```

Comment:
Could we use `GetAttrCallNode` here for `getAttrCall`? Then we would be a bit more sure that this is the method being called, and `arg = getAttrCall.getObject()` would also cover the `object=` case.

Reply:
> I also don't see how this would improve things. attr.(CfgNode).getNode() = getAttrCall.getNode() makes sure that the AttrRead and the DataFlowCall shares the underlying CFG node... isn't that enough?

It probably is, but the argument is by casing on the implemented `AttrRead`s and their shape. it might be nicer to ensure directly that `getAttrCall` is a call to the Python function `getattr`.

I will not push hard for this, though, the tests will probably make noise if it becomes a problem..